gateway/src/apicast/http_proxy.lua

```diff
@@ -82,14 +88,58 @@ local function absolute_url(uri)
 end
 
 local function forward_https_request(proxy_uri, uri, skip_https_connect)
-    -- This is needed to call ngx.req.get_body_data() below.
-    ngx.req.read_body()
+    local sock, ok, err
+    local chunksize = 32 * 1024
+    local body
+    local headers = req_headers()
+
+    local encoding = headers.transfer_encoding
+    if type(encoding) == "table" then
+        encoding = encoding[1]
+    end
 
-    local request = {
-        uri = uri,
-        method = ngx.req.get_method(),
-        headers = ngx.req.get_headers(0, true),
-        path = format('%s%s%s', ngx.var.uri, ngx.var.is_args, ngx.var.query_string or ''),
+    if encoding and encoding:lower() == "chunked" then
+        if http_ver() ~= 1.1 then
+            ngx.log(ngx.ERR, "bad http version")
+            ngx.exit(ngx.HTTP_BAD_REQUEST)
+         end
+         local expect = headers["Expect"]
+
+        if type(expect) == "table" then
+            expect = expect[1]
+        end
+
+        if expect and expect:lower() == "100-continue" then
+            ngx.status = 100
+            ok, err = ngx_send_headers()
+
+            if not ok then
+                ngx.log(ngx.ERR, "failed to send response header: " .. (err or "unknown"))
+            end
+
+            ok, err = ngx_flush(true)
+            if not ok then
+                ngx.log(ngx.ERR, "failed to flush response header: " .. (err or "unknown"))
+            end
+        end
+
+        -- The default ngx reader does not support chunked request
+        -- so we will need to get the raw request socket and manually
+        -- decode the chunked request
+        sock, err = ngx.req.socket(true)
+
+        if not sock then
+            if err == "no body" then
+                body = nil
+            else
+                return ngx.exit(ngx.HTTP_SERVICE_UNAVAILABLE)
+            end
+        else
+            body = chunked_reader(sock, chunksize)
```

Comment:
This method reads the entire body? In memory? What if body size is Gb length?

Reply:
it returns an iterator which can be used by lua-resty-http to read the body in chunk

Reply:
So, if it is an iterator, the request is not buffered. Then, APIcast will open a connection (or reuse existing one) as soon as there is some bytes to read (first chunk). Then, APIcast is not protecting upstream from Slowloris DDoS. Am I right?

Reply:
Yes, you are right, maybe we need a better solution, any ideas?